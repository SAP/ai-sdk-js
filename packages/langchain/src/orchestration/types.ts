import type {
  Prompt,
  TemplatingModuleConfig,
  OrchestrationModuleConfig as OrchestrationModuleConfigWithStringTemplating,
  ChatCompletionTool,
  StreamOptions
} from '@sap-ai-sdk/orchestration';
import type {
  BaseChatModelCallOptions,
  BindToolsInput
} from '@langchain/core/language_models/chat_models';
import type { CustomRequestConfig } from '@sap-ai-sdk/core';

/**
 * Tool type for LangChain Orchestration client.
 */
export type ChatOrchestrationToolType = ChatCompletionTool | BindToolsInput;

/**
 * Options for an orchestration call.
 */
export type OrchestrationCallOptions = Pick<
  BaseChatModelCallOptions,
  | 'stop'
  | 'signal'
  | 'timeout'
  | 'callbacks'
  | 'metadata'
  | 'runId'
  | 'runName'
  | 'tags'
> & {
  customRequestConfig?: CustomRequestConfig;
  strict?: boolean;
  tools?: ChatOrchestrationToolType[];
  promptIndex?: number;
  inputParams?: Prompt['inputParams'];
  streamOptions?: StreamOptions;
};

/**
 * Orchestration module configuration for LangChain.
 */
export type LangChainOrchestrationModuleConfig = Omit<
  OrchestrationModuleConfigWithStringTemplating,
  'templating'
> & {
  templating?: TemplatingModuleConfig;
};

/**
 * Orchestration module configuration for LangChain.
 * @deprecated Use `LangChainOrchestrationModuleConfig` instead.
 */
export type LangchainOrchestrationModuleConfig =
  LangChainOrchestrationModuleConfig;
