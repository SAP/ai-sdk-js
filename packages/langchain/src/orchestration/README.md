# LangChain Orchestration Client

LangChain orchestration client utilizes `@sap-ai-sdk/orchestration` package and provides an interface to interact with the Orchestration Service when using LangChain.

> **Note**: For installation and prerequisites of the package `@sap-ai-sdk/langchain`, refer to the [README](../../README.md).

### Table of Contents

- [Usage](#usage)
  - [Client Initialization](#client-initialization)
  - [Client Invocation](#client-invocation)
- [Local Testing](#local-testing)
- [Limitations](#limitations)

## Usage

The client complies with [LangChain's interface](https://js.langchain.com/docs/introduction).
Currently, streaming is not supported.

### Client Initialization

To initialize a client, four different configurations can be provided.
Orchestration configuration is mandatory and its structure is explained in our [`orchestration client`](https://github.com/SAP/ai-sdk-js/blob/main/packages/orchestration/README.md).
Optionally, it is possible to set [LangChain options](https://v03.api.js.langchain.com/types/_langchain_core.language_models_chat_models.BaseChatModelParams.html), a custom resource group, and a destination of the SAP AI Core service.

Below is an example for instantiating the orchestration client for LangChain:

```ts
import { OrchestrationClient } from '@sap-ai-sdk/langchain';
const config: OrchestrationModuleConfig = {
  llm: {
    model_name: 'gpt-4o'
  },
  templating: {
    template: [
      { role: 'user', content: 'Give me a long introduction of {{?subject}}' }
    ]
  }
};

const client = new OrchestrationClient(config);
```

#### Custom Destination

The `OrchestrationClient` can be initialized with a custom destination of the SAP AI Core service.
For example, use the following code to target a specific destination:

```ts
const client = new OrchestrationClient(
  orchestrationConfig,
  langchainOptions,
  deploymentConfig,
  { destinationName: 'my-destination' }
);
```

By default, the fetched destination is cached.
To disable caching, set the `useCache` parameter to `false` together with the `destinationName` parameter.

### Client Invocation

When invoking the client, pass a message history and, in most cases, input parameters for the template module.

```ts
const systemMessage = new SystemMessage('Be a helpful assisstant!');
const history = [systemMessage];
const response = await client.invoke(history, {
  inputParams: { subject: 'Paris' }
});
```

#### Resilience

To add resilience to the client, use LangChain's default options, especially `timeout` and `maxRetry`.

##### Timeout

By default, no timeout is set in the client.
To limit the maximum duration for the entire request, including retries, specify a timeout in milliseconds when using the `invoke` method:

```ts
const response = await client.invoke(messageHistory, { timeout: 10000 });
```

##### Retry

LangChain clients retry up to 6 times with exponential backoff by default.
To modify this behavior, set the `maxRetries` option during client initialization:

```ts
const client = new OrchestrationClient(orchestrationConfig, {
  maxRetries: 0
});
```

If the error is caused by input content filtering, the client will throw immediately without retrying.

## Local Testing

For local testing instructions, refer to this [section](https://github.com/SAP/ai-sdk-js/blob/main/README.md#local-testing).

## Limitations

Currently unsupported features are:

- Streaming
- Tool Calling
