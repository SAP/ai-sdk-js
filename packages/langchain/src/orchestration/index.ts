export * from './client.js';
export * from './orchestration-message.js';
export * from './orchestration-message-chunk.js';
export * from './types.js';
export * from './util.js';
