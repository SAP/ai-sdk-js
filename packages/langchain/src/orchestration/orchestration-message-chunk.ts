import { AIMessageChunk } from '@langchain/core/messages';
import type { AIMessageChunkFields } from '@langchain/core/messages';
import type { ModuleResultsStreaming } from '@sap-ai-sdk/orchestration';

/**
 * An AI Message Chunk containing module results and request ID.
 */
export class OrchestrationMessageChunk extends AIMessageChunk {
  module_results: ModuleResultsStreaming;
  request_id: string;
  constructor(
    fields: string | AIMessageChunkFields,
    module_results: ModuleResultsStreaming,
    request_id: string
  ) {
    super(fields);
    this.module_results = module_results;
    this.request_id = request_id;
  }
}
