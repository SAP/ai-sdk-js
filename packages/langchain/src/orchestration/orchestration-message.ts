import { AIMessage } from '@langchain/core/messages';
import type { AIMessageFields } from '@langchain/core/messages';
import type { ModuleResults } from '@sap-ai-sdk/orchestration';

/**
 * An AI Message containing module results and request ID.
 * @internal
 */
export class OrchestrationMessage extends AIMessage {
  module_results: ModuleResults;
  request_id: string;
  constructor(
    fields: string | AIMessageFields,
    module_results: ModuleResults,
    request_id: string
  ) {
    super(fields);
    this.module_results = module_results;
    this.request_id = request_id;
  }
}
