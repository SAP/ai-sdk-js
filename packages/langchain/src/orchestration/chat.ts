import { BaseChatModel } from '@langchain/core/language_models/chat_models';
import { OrchestrationClient as OrchestrationClientBase } from '@sap-ai-sdk/orchestration';
import { AsyncCaller } from '@langchain/core/utils/async_caller';
import { resilience } from '@sap-cloud-sdk/resilience';
import {
  isTemplate,
  mapLangchainMessagesToOrchestrationMessages,
  mapOutputToChatResult
} from './util.js';
import type { CustomRequestConfig } from '@sap-ai-sdk/core';
import type { OrchestrationMessageChunk } from './orchestration-message-chunk.js';
import type { ChatResult } from '@langchain/core/outputs';
import type { OrchestrationModuleConfig } from '@sap-ai-sdk/orchestration';
import type { BaseChatModelParams } from '@langchain/core/language_models/chat_models';
import type { ResourceGroupConfig } from '@sap-ai-sdk/ai-api';
import type { BaseMessage } from '@langchain/core/messages';
import type { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import type { OrchestrationCallOptions } from './types.js';
import type { HttpDestinationOrFetchOptions } from '@sap-cloud-sdk/connectivity';

// TODO: Update all docs

/**
 * LangChain chat client for Azure OpenAI consumption on SAP BTP.
 */
export class OrchestrationClient extends BaseChatModel<
  OrchestrationCallOptions,
  OrchestrationMessageChunk
> {
  constructor(
    // Omit streaming until supported
    public orchestrationConfig: Omit<OrchestrationModuleConfig, 'streaming'>,
    public langchainOptions: BaseChatModelParams = {},
    public deploymentConfig?: ResourceGroupConfig,
    public destination?: HttpDestinationOrFetchOptions
  ) {
    super(langchainOptions);
  }

  _llmType(): string {
    return 'orchestration';
  }

  /**
   * Decisions:
   * bind only supports ParsedCallOptions, we don't support arbitrary LLM options, only tool calls & default BaseLanguageModelCallOptions, e.g. stop ✅
   * this aligns with other vendors' client designs (e.g. openai, google) ✅
   * top of the array (array[array.length - 1]) contains the current message, everything before then is history.
   * Module results are part of our own message type, which extends AI Message to work with all other langchain functionality.
   *
   * For timeout, we need to apply our own middleware, it is not handled by langchain. ✅.
   */

  override async _generate(
    messages: BaseMessage[],
    options: typeof this.ParsedCallOptions,
    runManager?: CallbackManagerForLLMRun
  ): Promise<ChatResult> {
    let caller = this.caller;
    if (options.maxConcurrency) {
      const { maxConcurrency, maxRetries, onFailedAttempt } =
        this.langchainOptions;
      caller = new AsyncCaller({
        maxConcurrency: maxConcurrency ?? options.maxConcurrency,
        maxRetries,
        onFailedAttempt
      });
    }
    const res = await caller.callWithOptions(
      {
        signal: options.signal
      },
      () => {
        // consider this.tools & this.stop property, merge it with template orchestration config
        const { inputParams } = options;
        const mergedOrchestrationConfig =
          this.mergeOrchestrationConfig(options);
        const orchestrationClient = new OrchestrationClientBase(
          mergedOrchestrationConfig,
          this.deploymentConfig,
          this.destination
        );
        const messagesHistory =
          mapLangchainMessagesToOrchestrationMessages(messages);
        const customRequestConfig: CustomRequestConfig = {
          ...options.customRequestConfig,
          middleware: resilience({ timeout: options.timeout })
        };
        return orchestrationClient.chatCompletion(
          {
            // how to handle tools here? doesn't really exist as input in orchestration as message history
            // make template a call option, to merge it ??
            messagesHistory,
            inputParams
          },
          customRequestConfig
        );
      }
    );

    const content = res.getContent();

    // we currently do not support streaming
    await runManager?.handleLLMNewToken(
      typeof content === 'string' ? content : ''
    );

    return mapOutputToChatResult(res.data);
  }

  private mergeOrchestrationConfig(
    options: typeof this.ParsedCallOptions
  ): OrchestrationModuleConfig {
    const { tools = [], stop = [] } = options;
    return {
      ...this.orchestrationConfig,
      llm: {
        ...this.orchestrationConfig.llm,
        model_params: {
          ...this.orchestrationConfig.llm.model_params,
          ...(stop.length && {
            stop: [
              ...(this.orchestrationConfig.llm.model_params?.stop || []),
              ...stop
            ]
          })
        }
      },
      templating: {
        ...this.orchestrationConfig.templating,
        ...(isTemplate(this.orchestrationConfig.templating) &&
          tools.length && {
            tools: [
              ...(this.orchestrationConfig.templating.tools || []),
              ...tools
            ]
          })
      }
    };
  }
}
