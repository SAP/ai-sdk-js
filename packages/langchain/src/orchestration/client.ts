import { BaseChatModel } from '@langchain/core/language_models/chat_models';
import { OrchestrationClient as OrchestrationClientBase } from '@sap-ai-sdk/orchestration';
import {
  isTemplate,
  mapLangchainMessagesToOrchestrationMessages,
  mapOutputToChatResult
} from './util.js';
import type { BaseLanguageModelInput } from '@langchain/core/language_models/base';
import type { Runnable, RunnableLike } from '@langchain/core/runnables';
import type { OrchestrationMessageChunk } from './orchestration-message-chunk.js';
import type { ChatResult } from '@langchain/core/outputs';
import type { OrchestrationModuleConfig } from '@sap-ai-sdk/orchestration';
import type { BaseChatModelParams } from '@langchain/core/language_models/chat_models';
import type { ResourceGroupConfig } from '@sap-ai-sdk/ai-api';
import type { BaseMessage } from '@langchain/core/messages';
import type { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import type { OrchestrationCallOptions } from './types.js';
import type { HttpDestinationOrFetchOptions } from '@sap-cloud-sdk/connectivity';

/**
 * The Orchestration client.
 */
export class OrchestrationClient extends BaseChatModel<
  OrchestrationCallOptions,
  OrchestrationMessageChunk
> {
  constructor(
    // TODO: Omit streaming until supported
    public orchestrationConfig: Omit<OrchestrationModuleConfig, 'streaming'>,
    public langchainOptions: BaseChatModelParams = {},
    public deploymentConfig?: ResourceGroupConfig,
    public destination?: HttpDestinationOrFetchOptions
  ) {
    super(langchainOptions);
  }

  _llmType(): string {
    return 'orchestration';
  }

  /**
   * Create a new runnable sequence that runs each individual runnable in series,
   * piping the output of one runnable into another runnable or runnable-like.
   * @param coerceable - A runnable, function, or object whose values are functions or runnables.
   * @returns A new runnable sequence.
   */
  override pipe<NewRunOutput>(
    coerceable: RunnableLike<OrchestrationMessageChunk, NewRunOutput>
  ): Runnable<
    BaseLanguageModelInput,
    Exclude<NewRunOutput, Error>,
    OrchestrationCallOptions
  > {
    return super.pipe(coerceable) as Runnable<
      BaseLanguageModelInput,
      Exclude<NewRunOutput, Error>,
      OrchestrationCallOptions
    >;
  }

  override async _generate(
    messages: BaseMessage[],
    options: typeof this.ParsedCallOptions,
    runManager?: CallbackManagerForLLMRun
  ): Promise<ChatResult> {
    const res = await this.caller.callWithOptions(
      {
        signal: options.signal
      },
      () => {
        const { inputParams, customRequestConfig } = options;
        const mergedOrchestrationConfig =
          this.mergeOrchestrationConfig(options);
        const orchestrationClient = new OrchestrationClientBase(
          mergedOrchestrationConfig,
          this.deploymentConfig,
          this.destination
        );
        const messagesHistory =
          mapLangchainMessagesToOrchestrationMessages(messages);
        return orchestrationClient.chatCompletion(
          {
            messagesHistory,
            inputParams
          },
          customRequestConfig
        );
      }
    );

    const content = res.getContent();

    // TODO: Add streaming as soon as we support it
    await runManager?.handleLLMNewToken(
      typeof content === 'string' ? content : ''
    );

    return mapOutputToChatResult(res.data);
  }

  private mergeOrchestrationConfig(
    options: typeof this.ParsedCallOptions
  ): OrchestrationModuleConfig {
    const { tools = [], stop = [] } = options;
    return {
      ...this.orchestrationConfig,
      llm: {
        ...this.orchestrationConfig.llm,
        model_params: {
          ...this.orchestrationConfig.llm.model_params,
          ...(stop.length && {
            stop: [
              ...(this.orchestrationConfig.llm.model_params?.stop || []),
              ...stop
            ]
          })
        }
      },
      templating: {
        ...this.orchestrationConfig.templating,
        ...(isTemplate(this.orchestrationConfig.templating) &&
          tools.length && {
            tools: [
              ...(this.orchestrationConfig.templating.tools || []),
              ...tools
            ]
          })
      }
    };
  }
}
