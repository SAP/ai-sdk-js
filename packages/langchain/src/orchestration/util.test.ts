import {
  AIMessage,
  FunctionMessage,
  AIMessageChunk,
  HumanMessage,
  SystemMessage,
  ToolMessage
} from '@langchain/core/messages';
import { OrchestrationStreamChunkResponse } from '@sap-ai-sdk/orchestration';
import { jest } from '@jest/globals';
import {
  mapLangChainMessagesToOrchestrationMessages,
  mapOutputToChatResult,
  setFinishReason,
  setTokenUsage,
  computeTokenIndices,
  mapOrchestrationChunkToLangChainMessageChunk
} from './util.js';
import type { OrchestrationMessage } from './orchestration-message.js';
import type { ToolCallChunk } from '@langchain/core/messages/tool';
import type {
  CompletionPostResponse,
  MessageToolCall,
  TokenUsage,
  ToolCallChunk as OrchestrationToolCallChunk,
  CompletionPostResponseStreaming
} from '@sap-ai-sdk/orchestration';

describe('mapLangChainMessagesToOrchestrationMessages', () => {
  it('should map an array of LangChain messages to Orchestration messages', () => {
    const langchainMessages = [
      new SystemMessage('System message content'),
      new HumanMessage('Human message content'),
      new AIMessage('AI message content'),
      new ToolMessage('Tool message content', 'tool_call_id')
    ];

    const result =
      mapLangChainMessagesToOrchestrationMessages(langchainMessages);

    expect(result).toEqual([
      { role: 'system', content: 'System message content' },
      { role: 'user', content: 'Human message content' },
      { role: 'assistant', content: 'AI message content' },
      {
        role: 'tool',
        content: 'Tool message content',
        tool_call_id: 'tool_call_id'
      }
    ]);
  });

  it('should throw error for unsupported message types', () => {
    const langchainMessages = [
      new FunctionMessage('Function message content', 'function_name')
    ];

    expect(() =>
      mapLangChainMessagesToOrchestrationMessages(langchainMessages)
    ).toThrow('Unsupported message type: function');
  });
});

describe('mapBaseMessageToChatMessage', () => {
  it('should map HumanMessage to ChatMessage with user role', () => {
    const humanMessage = new HumanMessage('Human message content');

    // Since mapBaseMessageToChatMessage is internal, we'll test it through mapLangChainMessagesToOrchestrationMessages
    const result = mapLangChainMessagesToOrchestrationMessages([humanMessage]);

    expect(result[0]).toEqual({
      role: 'user',
      content: 'Human message content'
    });
  });

  it('should map HumanMessage with `image_url` to ChatMessage with user role', () => {
    const humanMessage = new HumanMessage({
      content: [
        { type: 'text', text: 'Human message content' },
        {
          type: 'image_url',
          image_url: 'https://example.com/image.jpg'
        }
      ]
    });

    const result = mapLangChainMessagesToOrchestrationMessages([humanMessage]);

    expect(result[0]).toEqual({
      role: 'user',
      content: [
        { type: 'text', text: 'Human message content' },
        {
          type: 'image_url',
          image_url: { url: 'https://example.com/image.jpg' }
        }
      ]
    });
  });

  it('should map SystemMessage to ChatMessage with system role', () => {
    const systemMessage = new SystemMessage('System message content');

    const result = mapLangChainMessagesToOrchestrationMessages([systemMessage]);

    expect(result[0]).toEqual({
      role: 'system',
      content: 'System message content'
    });
  });

  it('should map AIMessage to ChatMessage with assistant role', () => {
    const aiMessage = new AIMessage('AI message content');

    const result = mapLangChainMessagesToOrchestrationMessages([aiMessage]);

    expect(result[0]).toEqual({
      role: 'assistant',
      content: 'AI message content'
    });
  });

  it('should map ToolMessage to ChatMessage with tool role', () => {
    const toolMessage = new ToolMessage('Tool message content', 'tool_call_id');
    const result = mapLangChainMessagesToOrchestrationMessages([toolMessage]);
    expect(result[0]).toEqual({
      role: 'tool',
      content: 'Tool message content',
      tool_call_id: 'tool_call_id'
    });
  });

  it('should throw error when mapping SystemMessage with unsupported content type like `image_url`', () => {
    const systemMessage = new SystemMessage({
      content: [
        { type: 'text', text: 'System text' },
        {
          type: 'image_url',
          image_url: { url: 'https://example.com/image.jpg' }
        }
      ]
    });

    expect(() =>
      mapLangChainMessagesToOrchestrationMessages([systemMessage])
    ).toThrow(
      'The content type of system message can only be "text" in the Orchestration Client.'
    );
  });

  it('should throw error when mapping ToolMessage with unsupported content type like `image_url`', () => {
    const toolMessage = new ToolMessage(
      {
        content: [
          { type: 'text', text: 'System text' },
          {
            type: 'image_url',
            image_url: { url: 'https://example.com/image.jpg' }
          }
        ]
      },
      'tool_call_id'
    );

    expect(() =>
      mapLangChainMessagesToOrchestrationMessages([toolMessage])
    ).toThrow(
      'The content type of tool message can only be "text" in the Orchestration Client.'
    );
  });
});

describe('mapOutputToChatResult', () => {
  it('should map CompletionPostResponse to ChatResult', () => {
    const completionResponse: CompletionPostResponse = {
      orchestration_result: {
        id: 'test-id',
        object: 'chat.completion',
        created: 1634840000,
        model: 'test-model',
        choices: [
          {
            message: {
              content: 'Test content',
              role: 'assistant'
            },
            finish_reason: 'stop',
            index: 0
          }
        ],
        usage: {
          completion_tokens: 10,
          prompt_tokens: 20,
          total_tokens: 30
        }
      },
      request_id: 'req-123',
      module_results: {}
    };

    const result = mapOutputToChatResult(completionResponse);

    expect(result.generations).toHaveLength(1);
    expect(result.generations[0].text).toBe('Test content');
    expect(result.generations[0].message.content).toBe('Test content');
    expect(result.generations[0].generationInfo).toEqual({
      finish_reason: 'stop',
      index: 0,
      function_call: undefined,
      tool_calls: undefined
    });
    expect(result.llmOutput).toEqual({
      created: 1634840000,
      id: 'test-id',
      model: 'test-model',
      object: 'chat.completion',
      tokenUsage: {
        completionTokens: 10,
        promptTokens: 20,
        totalTokens: 30
      }
    });
  });

  it('should map tool_calls correctly', () => {
    const toolCallData: MessageToolCall = {
      id: 'call-123',
      type: 'function',
      function: {
        name: 'test_function',
        arguments: '{"arg1":"value1"}'
      }
    };

    const completionResponse: CompletionPostResponse = {
      orchestration_result: {
        id: 'test-id',
        object: 'chat.completion',
        created: 1634840000,
        model: 'test-model',
        choices: [
          {
            index: 0,
            message: {
              content: 'Test content',
              role: 'assistant',
              tool_calls: [toolCallData]
            },
            finish_reason: 'tool_calls'
          }
        ],
        usage: {
          completion_tokens: 10,
          prompt_tokens: 20,
          total_tokens: 30
        }
      },
      request_id: 'req-123',
      module_results: {}
    };

    const result = mapOutputToChatResult(completionResponse);

    expect(
      (result.generations[0].message as OrchestrationMessage).tool_calls
    ).toEqual([
      {
        id: 'call-123',
        name: 'test_function',
        args: { arg1: 'value1' },
        type: 'tool_call'
      }
    ]);
  });
});

describe('setFinishReason', () => {
  it('should set finish reason on message chunk when provided', () => {
    const messageChunk = new AIMessageChunk({ content: 'Test content' });

    setFinishReason(messageChunk, 'stop');

    expect(messageChunk.response_metadata.finish_reason).toBe('stop');
  });

  it('should not modify response_metadata when finish reason is falsy', () => {
    const messageChunk = new AIMessageChunk({ content: 'Test content' });
    const originalMetadata = { ...messageChunk.response_metadata };

    setFinishReason(messageChunk, '');

    expect(messageChunk.response_metadata).toEqual(originalMetadata);
  });
});

describe('setTokenUsage', () => {
  it('should set token usage metadata when provided', () => {
    const messageChunk = new AIMessageChunk({ content: 'Test content' });
    const tokenUsage: TokenUsage = {
      completion_tokens: 10,
      prompt_tokens: 20,
      total_tokens: 30
    };

    setTokenUsage(messageChunk, tokenUsage);

    expect(messageChunk.usage_metadata).toEqual({
      input_tokens: 20,
      output_tokens: 10,
      total_tokens: 30
    });
    expect(messageChunk.response_metadata.token_usage).toEqual(tokenUsage);
  });

  it('should not modify message chunk when token usage is undefined', () => {
    const messageChunk = new AIMessageChunk({
      content: 'Test content',
      response_metadata: { finish_reason: 'stop' }
    });

    setTokenUsage(messageChunk, undefined);

    expect(messageChunk.usage_metadata).toBeUndefined();
    expect(messageChunk.response_metadata.token_usage).toBeUndefined();
    expect(messageChunk.response_metadata.finish_reason).toBe('stop');
  });
});

describe('computeTokenIndices', () => {
  it('should compute token indices with choice index from chunk', () => {
    const mockChunk = new OrchestrationStreamChunkResponse({
      request_id: 'req-123',
      orchestration_result: {
        id: 'test-id',
        object: 'chat.completion.chunk',
        created: 1634840000,
        model: 'test-model',
        system_fingerprint: 'fp_123',
        choices: [
          {
            index: 0,
            delta: { content: 'Test' }
          }
        ]
      }
    });

    const result = computeTokenIndices(mockChunk);

    expect(result).toEqual({
      prompt: 0,
      completion: 0
    });
  });
});

describe('mapOrchestrationChunkToLangChainMessageChunk', () => {
  function createMockChunk(
    content?: string,
    toolCallChunks?: OrchestrationToolCallChunk[],
    finishReason?: string,
    tokenUsage?: {
      completion_tokens: number;
      prompt_tokens: number;
      total_tokens: number;
    }
  ): OrchestrationStreamChunkResponse {
    const mockData: CompletionPostResponseStreaming = {
      request_id: 'req-123',
      module_results: {
        llm: {
          id: 'test-id',
          object: 'chat.completion.chunk',
          created: 1634840000,
          model: 'test-model',
          system_fingerprint: 'fp_123',
          choices: [
            {
              index: 0,
              delta: {
                role: 'assistant',
                content: content || '',
                tool_calls: toolCallChunks
              },
              finish_reason: finishReason || ''
            }
          ],
          usage: tokenUsage
        }
      },
      orchestration_result: {
        id: 'test-id',
        object: 'chat.completion.chunk',
        created: 1634840000,
        model: 'test-model',
        system_fingerprint: 'fp_123',
        choices: [
          {
            index: 0,
            delta: {
              role: 'assistant',
              content: content || '',
              tool_calls: toolCallChunks
            },
            finish_reason: finishReason || ''
          }
        ],
        usage: tokenUsage
      }
    };

    const mockChunk = new OrchestrationStreamChunkResponse(mockData);
    jest.spyOn(mockChunk, 'getDeltaContent').mockReturnValue(content);
    jest.spyOn(mockChunk, 'getDeltaToolCalls').mockReturnValue(toolCallChunks);

    return mockChunk;
  }

  it('should map a chunk with content to AIMessageChunk', () => {
    const mockChunk = createMockChunk('Test content');
    const result = mapOrchestrationChunkToLangChainMessageChunk(mockChunk);

    expect(result).toBeInstanceOf(AIMessageChunk);
    expect(result.content).toBe('Test content');
    expect(result.additional_kwargs).toEqual({
      module_results: mockChunk.data.module_results,
      request_id: 'req-123'
    });
    expect(result.tool_call_chunks).toEqual([]);
    expect(result).toMatchSnapshot('AIMessageChunk with content');
  });

  it('should map a chunk with tool call chunks to AIMessageChunk', () => {
    const toolCallChunk: OrchestrationToolCallChunk = {
      id: 'call-123',
      index: 0,
      type: 'function',
      function: {
        name: 'test_function',
        arguments: ''
      }
    };
    const mockChunk = createMockChunk('', [toolCallChunk]);
    const result = mapOrchestrationChunkToLangChainMessageChunk(mockChunk);

    expect(result).toBeInstanceOf(AIMessageChunk);
    expect(result.content).toBe('');
    expect(result.tool_call_chunks).toHaveLength(1);
    const expectedToolCallChunk: ToolCallChunk = {
      id: 'call-123',
      index: 0,
      name: 'test_function',
      args: '',
      type: 'tool_call_chunk'
    };
    expect(result.tool_call_chunks?.[0]).toEqual(expectedToolCallChunk);
    expect(result).toMatchSnapshot('AIMessageChunk with tool call chunks');
  });
});
