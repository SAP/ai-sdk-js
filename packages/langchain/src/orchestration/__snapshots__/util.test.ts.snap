// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapOrchestrationChunkToLangChainMessageChunk should map a chunk with content to AIMessageChunk: AIMessageChunk with content 1`] = `
{
  "id": [
    "langchain_core",
    "messages",
    "AIMessageChunk",
  ],
  "kwargs": {
    "additional_kwargs": {
      "module_results": {
        "llm": {
          "choices": [
            {
              "delta": {
                "content": "Test content",
                "role": "assistant",
                "tool_calls": undefined,
              },
              "finish_reason": "",
              "index": 0,
            },
          ],
          "created": 1634840000,
          "id": "test-id",
          "model": "test-model",
          "object": "chat.completion.chunk",
          "system_fingerprint": "fp_123",
          "usage": undefined,
        },
      },
    },
    "content": "Test content",
    "invalid_tool_calls": [],
    "response_metadata": {},
    "tool_call_chunks": [],
    "tool_calls": [],
    "usage_metadata": undefined,
  },
  "lc": 1,
  "type": "constructor",
}
`;

exports[`mapOrchestrationChunkToLangChainMessageChunk should map a chunk with tool call chunks to AIMessageChunk: AIMessageChunk with tool call chunks 1`] = `
{
  "id": [
    "langchain_core",
    "messages",
    "AIMessageChunk",
  ],
  "kwargs": {
    "additional_kwargs": {
      "module_results": {
        "llm": {
          "choices": [
            {
              "delta": {
                "content": "",
                "role": "assistant",
                "tool_calls": [
                  {
                    "function": {
                      "arguments": "",
                      "name": "test_function",
                    },
                    "id": "call-123",
                    "index": 0,
                    "type": "function",
                  },
                ],
              },
              "finish_reason": "",
              "index": 0,
            },
          ],
          "created": 1634840000,
          "id": "test-id",
          "model": "test-model",
          "object": "chat.completion.chunk",
          "system_fingerprint": "fp_123",
          "usage": undefined,
        },
      },
    },
    "content": "",
    "invalid_tool_calls": [],
    "response_metadata": {},
    "tool_call_chunks": [
      {
        "args": "",
        "id": "call-123",
        "index": 0,
        "name": "test_function",
        "type": "tool_call_chunk",
      },
    ],
    "tool_calls": [
      {
        "args": {},
        "id": "call-123",
        "name": "test_function",
        "type": "tool_call",
      },
    ],
    "usage_metadata": undefined,
  },
  "lc": 1,
  "type": "constructor",
}
`;
